### DISCUSSION TITLE
PID contoller interface and software PID controller

### DISCUSSION THREAD

**Discussion by qku** [Ulm-IQO/qudi-core#52] [Ideas]
I am currently planning to implement more PID-control related hardware and was thinking about the PID toolchain in general. In my now [merged PR on iqo-modules](https://github.com/Ulm-IQO/qudi-iqo-modules/pull/29), @timoML suggested to unify the `PIDControllerInterface` with the `ProcessSetpointInterface`. Do you or @Neverhorst or someone else have more thoughts on this? I was thinking that it would be good in general to keep a `PIDControllerInterface` for dedicated PID hardware, e.g. a RedPitaya.

For software PID control, there is a `SoftPIDController` currently living in the `logic` folder even though it is subclassing `PIDControllerInterface`. Isn't this actually an interfuse that connects to process setpoint and value hardware?

**Comment by LukePiewalker** [comment]
Dear @qku,

could it be that the unification should have been aimed to use the 'ProcessSetpointInterface' inside the 'PIDControllerInterface'? When I look at the 'PIDControllerInterface', it basically implements many methods that the 'ProcessControlInterface' has to offer, like setpoints and process values. The only thing I don't see directly would be something like "enable" and "disable" PID, but this could be feasible via the "activity_state" property? However, with the 'ProcessControlInterface' it's not really clear what should happen when setting the "activity_state". Intuitively, I would say that setting "active" activates setting and retrieving process and setpoint values, and setting "unactive" releases hardware resources, for example.

As for the "SoftPIDController", I assume it is still there since interfuses had been in the logic folder in the old core while we have now moved them to hardware.
